src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,431 @@
+                              op->extents,
+                              op->condition,
+                              std::move(body),
+                              op->new_expr,
+                              op->free_function);
+    }
+
+    Stmt visit(const ProducerConsumer *op) override {
+        // Usually we allocate the mutex buffer at the Allocate node,
+        // but outputs don't have Allocate. For those we allocate the mutex
+        // buffer at the producer node.
+
+        if (!op->is_producer) {
+            // This is a consumer.
+            return IRMutator::visit(op);
+        }
+
+        // Find the corresponding output.
+        auto func_it = env.find(op->name);
+        if (func_it == env.end()) {
+            // Not an output.
+            return IRMutator::visit(op);
+        }
+        Func f = Func(func_it->second);
+        if (f.output_buffers().size() == 0) {
+            // Not an output.
+            return IRMutator::visit(op);
+        }
+
+        set<string> store_names;
+        for (auto buffer : f.output_buffers()) {
+            store_names.insert(buffer.name());
+        }
+
+        FindStoreInAtomicMutex finder(store_names);
+        op->body.accept(&finder);
+        if (!finder.found) {
+            // No Atomic node that requires mutex lock from this node inside.
+            return IRMutator::visit(op);
+        }
+
+        if (allocated_mutexes.find(finder.mutex_name) != allocated_mutexes.end()) {
+            // We've already allocated a mutex.
+            return IRMutator::visit(op);
+        }
+
+        allocated_mutexes.insert(finder.mutex_name);
+
+        // We assume all output buffers in a Tuple have the same extent.
+        OutputImageParam output_buffer = f.output_buffers()[0];
+        Expr extent = Expr(1);
+        for (int i = 0; i < output_buffer.dimensions(); i++) {
+            extent = extent * output_buffer.dim(i).extent();
+        }
+        Stmt body = mutate(op->body);
+        body = allocate_mutex(finder.mutex_name, extent, body);
+        return ProducerConsumer::make(op->name, op->is_producer, std::move(body));
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (op->mutex_name.empty()) {
+            return IRMutator::visit(op);
+        }
+
+        // Lock the mutexes using the indices from the Store nodes inside the body.
+        FindStoreIndex find;
+        op->body.accept(&find);
+
+        Stmt body = op->body;
+
+        Expr index = find.index;
+        Expr index_let; // If defined, represents the value of the lifted let binding.
+        if (!index.defined()) {
+            // Scalar output.
+            index = Expr(0);
+        } else {
+            // Lift the index outside of the atomic node.
+            string name = unique_name('t');
+            index_let = index;
+            index = Variable::make(index.type(), name);
+            body = ReplaceStoreIndexWithVar(index).mutate(body);
+        }
```

Comment:
All of this code assumes that all stores inside the atomic node are to the Func that the atomic node is guarding. What happens if additional stores get injected here, e.g. due to tracing or profiling or other future instrumentation? Should there be some sort of name check against the name on the atomic node?